### PR TITLE
Refactor release process for Github and Staging

### PR DESCRIPTION
**Pull request by randomvariable**
Replace use of sed with gojq to modify templates, by copying manifests to _artifacts and outputting to out/*-components.yaml, preventing source files being modified in place.

Switch to using Docker build cache to speed up builds, which should reduce CloudBuild times significantly.

CloudBuild modified such that manifests will be generated and uploaded to GCS on execution. This will allow advanced users to be able to test manifests directly from staging for testing prior to release.

Finally, targets are added to create a release in Github using the Github CLI.

Signed-off-by: Naadir Jeewa <jeewan@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by MarcusNoble** [comment]
This all looks good to me. Only thing I'd suggest is if possible are we able to break the Makefile down any more? It's becoming quite an intimidating file.

File: docs/book/src/development/releasing.md [hunk 1]
```diff
@@ -4,25 +4,26 @@
 
 1. Make sure your repo is clean by git's standards
 2. If this is a new minor release, create a new release branch and push to github, otherwise switch to it, for example `release-0.2`
-3. Run `make release-notes` to gather changes since the last revision. If you need to specify a specific tag to look for changes
```
Comment (sedefsavas) [inline docs/book/src/development/releasing.md:7]: Do we want to remove this?

File: Makefile [hunk 2]
```diff
@@ -326,82 +338,58 @@ docker-push-%:
...
+.PHONY: docker-push-manifest
+docker-push-manifest:
+	docker manifest create --amend $(CONTROLLER_IMG):$(TAG) $(shell echo $(ALL_ARCH) | sed -e "s~[^ ]*~$(CONTROLLER_IMG)\-&:$(TAG)~g")
+	@for arch in $(ALL_ARCH); do docker manifest annotate --arch $${arch} ${CONTROLLER_IMG}:${TAG} ${CONTROLLER_IMG}-$${arch}:${TAG}; done
+	docker manifest push --purge ${CONTROLLER_IMG}:${TAG}
+
+.PHONY: staging-manifests
+staging-manifests:
+	$(MAKE) $(RELEASE_DIR)/$(CORE_MANIFEST_FILE).yaml PULL_POLICY=IfNotPresent TAG=$(RELEASE_ALIAS_TAG)
+	$(MAKE) $(RELEASE_DIR)/$(EKS_BOOTSTRAP_MANIFEST_FILE).yaml PULL_POLICY=IfNotPresent TAG=$(RELEASE_ALIAS_TAG)
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply Makefile:362]: I had rebased on the broken main branch. Rebased again, should be ok.

File: Makefile [hunk 3]
```diff
@@ -415,32 +403,123 @@ release-binary: $(RELEASE_DIR) versions.mk
...
+	$(MAKE) compiled-manifest \
+	PROVIDER=$(EKS_CONTROLPLANE_MANIFEST_FILE) \
+	OLD_IMG=$(EKS_CONTROLPLANE_CONTROLLER_ORIGINAL_IMG) \
+	MANIFEST_IMG=$(EKS_CONTROLPLANE_CONTROLLER_IMG) \
+	CONTROLLER_NAME=$(EKS_CONTROLPLANE_CONTROLLER_NAME) \
+	PROVIDER_CONFIG_DIR=$(EKS_CONTROLPLANE_CONFIG_DIR) \
+	NAMESPACE=$(EKS_CONTROLPLANE_NAMESPACE)
+
+.PHONY: compiled-manifest
+compiled-manifest: $(RELEASE_DIR) $(KUSTOMIZE)
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply Makefile:484]: wth

[the remark this answers is not in the document]
Comment (randomvariable) [reply Makefile:484]: How about now?

**Comment by randomvariable** [comment]
> This all looks good to me. Only thing I'd suggest is if possible are we able to break the Makefile down any more? It's becoming quite an intimidating file.

I think as a followup, probably start moving the make targets for each controller into their own directories.

File: docs/book/src/development/releasing.md (hunk 1 above)
Comment (randomvariable) [reply docs/book/src/development/releasing.md:7]: Yup. `make release` will output it in out/CHANGELOG.md, and have amended this doc to reference it.

File: Makefile [hunk 4]
```diff
@@ -326,82 +338,58 @@ docker-push-%:
...
-	cp metadata.yaml $(RELEASE_DIR)/metadata.yaml
 
 .PHONY: release-manifests
-release-manifests: $(RELEASE_DIR) $(KUSTOMIZE) ## Builds the manifests to publish with a release
-	$(KUSTOMIZE) build config > $(RELEASE_DIR)/infrastructure-components.yaml
-	$(KUSTOMIZE) build bootstrap/eks/config > $(RELEASE_DIR)/eks-bootstrap-components.yaml
-	$(KUSTOMIZE) build controlplane/eks/config > $(RELEASE_DIR)/eks-controlplane-components.yaml
+release-manifests:
+	$(MAKE) $(RELEASE_DIR)/$(CORE_MANIFEST_FILE).yaml TAG=$(RELEASE_TAG) PULL_POLICY=IfNotPresent
+	$(MAKE) $(RELEASE_DIR)/$(EKS_BOOTSTRAP_MANIFEST_FILE).yaml TAG=$(RELEASE_TAG) PULL_POLICY=IfNotPresent
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply Makefile:389]: Ah, it doesn't work on a newer version of kustomize.

File: Makefile [hunk 5]
```diff
@@ -415,32 +403,121 @@ release-binary: $(RELEASE_DIR) versions.mk
...
+	$(MAKE) staging-manifests
+	$(MAKE) upload-staging-artifacts
 
-RELEASE_ALIAS_TAG=$(PULL_BASE_REF)
+.PHONY: upload-staging-artifacts
+upload-staging-artifacts: ## Upload release artifacts to the staging bucket
+	gsutil cp $(RELEASE_DIR)/* gs://$(BUCKET)/components/$(RELEASE_ALIAS_TAG)
+
+.PHONY: create-gh-release
+create-gh-release:$(GH) ## Create release on Github
```
Comment (sedefsavas) [inline Makefile:426]: I observe this error on Mac:

```
 tar -xvf share/gh/gh.tar.gz gh_1.2.0_macOS_amd64/bin/gh  --strip-components 1 --directory bin
x gh_1.2.0_macOS_amd64/bin/gh
tar: --strip-components: Not found in archive

```

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (randomvariable) [reply Makefile:389]: Should be fixed now.

File: Makefile (hunk 5 above)
Comment (randomvariable) [reply Makefile:426]: Switched to gtar, and added a check to make sure GNU Tar is installed.

File: Makefile [hunk 6]
```diff
@@ -326,82 +338,58 @@ docker-push-%:
...
 
-.PHONY: release
-release: clean-release  ## Builds and push container images using the latest git tag for the commit.
+.PHONY: check-release-tag
+check-release-tag:
 	@if [ -z "${RELEASE_TAG}" ]; then echo "RELEASE_TAG is not set"; exit 1; fi
 	@if ! [ -z "$$(git status --porcelain)" ]; then echo "Your local git repository contains uncommitted changes, use git clean before proceeding."; exit 1; fi
+
+.PHONY: release
+release: clean-release check-release-tag $(RELEASE_DIR)  ## Builds and push container images using the latest git tag for the commit.
```
Comment (sedefsavas) [inline Makefile:378]:
```suggestion
release: clean-release check-release-tag release-notes $(RELEASE_DIR)  ## Builds and push container images using the latest git tag for the commit.
```

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (sedefsavas) [reply Makefile:362]: ` make release` still shows the same error.

File: Makefile (hunk 5 above)
Comment (sedefsavas) [reply Makefile:426]: Thought I pulled the updated one, will rebase and try again.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (sedefsavas) [reply Makefile:362]: Please ignore for now I guess need to rebase again.

File: Makefile [hunk 7]
```diff
@@ -326,82 +338,58 @@ docker-push-%:
...
-	$(MAKE) set-manifest-image \
-		MANIFEST_IMG=$(PROD_REGISTRY)/$(EKS_BOOTSTRAP_IMAGE_NAME) MANIFEST_TAG=$(RELEASE_TAG) \
-		TARGET_RESOURCE="./bootstrap/eks/config/manager/manager_image_patch.yaml"
-	# Set manifest image for EKS controlplane provider to the production bucket.
-	$(MAKE) set-manifest-image \
-		MANIFEST_IMG=$(PROD_REGISTRY)/$(EKS_CONTROLPLANE_IMAGE_NAME) MANIFEST_TAG=$(RELEASE_TAG) \
-		TARGET_RESOURCE="./controlplane/eks/config/manager/manager_image_patch.yaml"
-	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./config/manager/manager_pull_policy.yaml"
-	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./bootstrap/eks/config/manager/manager_pull_policy.yaml"
-	$(MAKE) set-manifest-pull-policy PULL_POLICY=IfNotPresent TARGET_RESOURCE="./controlplane/eks/config/manager/manager_pull_policy.yaml"
```
Comment (sedefsavas) [inline Makefile:394]: New manifest pull policies under out/ is `Always`.
There must be a problem with patching.

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (sedefsavas) [reply Makefile:484]: It works now.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (sedefsavas) [reply Makefile:362]: yes this is fixed now.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (sedefsavas) [reply Makefile:389]: Fixed.

File: Makefile [hunk 8]
```diff
@@ -415,32 +403,121 @@ release-binary: $(RELEASE_DIR) versions.mk
...
+	$(MAKE) upload-staging-artifacts
 
-RELEASE_ALIAS_TAG=$(PULL_BASE_REF)
+.PHONY: upload-staging-artifacts
+upload-staging-artifacts: ## Upload release artifacts to the staging bucket
+	gsutil cp $(RELEASE_DIR)/* gs://$(BUCKET)/components/$(RELEASE_ALIAS_TAG)
+
+.PHONY: create-gh-release
+create-gh-release:$(GH) ## Create release on Github
+	$(GH) release create $(VERSION) -d -F $(RELEASE_DIR)/README.md -t $(VERSION) -R $(GH_REPO)
```
Comment (sedefsavas) [inline Makefile:427]:
```suggestion
	$(GH) release create $(VERSION) -d -F $(RELEASE_DIR)/CHANGELOG.md -t $(VERSION) -R $(GH_REPO)
```

File: docs/book/src/development/releasing.md [hunk 9]
```diff
@@ -4,25 +4,26 @@
...
-   since, use `make release-notes ARGS="--from <tag>"` Pay close attention to the `## :question: Sort these by hand` section, as it contains items that need to be manually sorted.
-4. Tag the repository and push the tag `git tag -s -m $VERSION $VERSION`
-5. Create a draft release in github and associate it with the tag that was just created, copying the generated release notes into
-   the draft.
-6. Checkout the tag you've just created and make sure git is in a clean state
-7. Run `make release`
-8. Attach the files to the drafted release:
+3. Tag the repository and push the tag `git tag -s -m $VERSION $VERSION`
+4. Set an environment variable PREVIOUS_VERSION to the last release tag.
+5. Create a draft release in github and associate it with the tag that was just created, copying the generated release notes
```
Comment (sedefsavas) [inline docs/book/src/development/releasing.md:9]: What is the difference of this with `make create-gh-release`?

File: Makefile (hunk 5 above)
Comment (sedefsavas) [reply Makefile:426]: Yep works.

File: docs/book/src/development/releasing.md (hunk 9 above)
Comment (randomvariable) [reply docs/book/src/development/releasing.md]: None, cleaned up.

File: Makefile (hunk 8 above)
Comment (randomvariable) [reply Makefile]: Done.

File: Makefile (hunk 6 above)
Comment (randomvariable) [reply Makefile:378]: added as a step in the target.

File: Makefile (hunk 7 above)
Comment (randomvariable) [reply Makefile:394]: sorted, i think.

File: Makefile (hunk 5 above)
Comment (randomvariable) [reply Makefile:427]: > Thought I pulled the updated one, will rebase and try again.

Had updated after you commented. :)

File: Makefile (hunk 7 above)
Comment (sedefsavas) [reply Makefile:394]: eks-bootstrap still has `Always`

```
        image: gcr.io/k8s-staging-cluster-api-aws/eks-bootstrap-controller:v0.6.1
        imagePullPolicy: Always
```

Rest is fixed.

**Comment by sedefsavas** [comment]
@randomvariable Thanks for the sed change! CAPI would benefit from this too.

This LGTM but do we want to include this in v0.6.2 release?

**Comment by randomvariable** [comment]
@sedefsavas don't need it in v0.6.2 at this stage.

**Comment by randomvariable** [comment]
/assign @ncdc 

tl;dr, finally stopped all the config files being rewritten.

**Review by richardcase — APPROVED** [review]
/lgtm

**Comment by randomvariable** [comment]
I've rebased this @richardcase
